File: src/features/ElementsBrowser/utils/savedSearchFilterDto.ts
import { ElementFilter } from "../../../api/savedSearches/dto/elementFilter";
import { SavedSearchFilterDto } from "../../../api/savedSearches/dto/savedSearchFilterDto";

export function toSavedSearchFilterDtos(
	filters: ElementFilter[],
): SavedSearchFilterDto[] {
	return filters.map((filter, index) => ({ index, filter }));
}

export function fromSavedSearchFilterDtos(
	filterDtos: SavedSearchFilterDto[],
): ElementFilter[] {
	return [...filterDtos]
		.sort((a, b) => a.index - b.index)
		.map(dto => dto.filter);
}

File: src-tauri/src/saved_searches/dto/saved_search_filter_dto.rs
use serde::{Deserialize, Serialize};

use crate::saved_searches::entities::saved_search_filter::{ElementFilter, SavedSearchFilter};

#[derive(Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct SavedSearchFilterDto {
    pub index: i64,
    pub filter: ElementFilter,
}

impl From<SavedSearchFilter> for SavedSearchFilterDto {
    fn from(filter: SavedSearchFilter) -> Self {
        SavedSearchFilterDto {
            index: filter.index,
            filter: filter.filter,
        }
    }
}

impl From<SavedSearchFilterDto> for SavedSearchFilter {
    fn from(dto: SavedSearchFilterDto) -> Self {
        SavedSearchFilter {
            index: dto.index,
            filter: dto.filter,
        }
    }
}
